File: app/src/main/java/me/knighthat/interactivedeck/connection/wireless/WirelessController.java
package me.knighthat.interactivedeck.connection.wireless;

import android.util.Log;

import androidx.annotation.IntRange;

import com.google.android.material.textfield.TextInputEditText;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.net.Socket;

import me.knighthat.interactivedeck.activity.DefaultActivity;

public class WirelessController extends Thread {

    public static @Nullable Thread THREAD;
    public static @Nullable Socket SOCKET;

    private final @NotNull String address;
    private final @IntRange ( from = 0x1, to = 0xffff ) int port;

    public WirelessController( @NotNull String address, @IntRange ( from = 0x1, to = 0xffff ) int port ) {
        this.address = address;
        this.port = port;
    }

    public static void connect( @NotNull TextInputEditText ipInput, @NotNull TextInputEditText portInput ) {
        if (ipInput.getText() == null || portInput.getText() == null)
            return;

        String ipAddr = ipInput.getText().toString();
        String portStr = portInput.getText().toString();
        int port = Integer.parseInt(portStr);

        if (port < 0x1 || port > 0xffff) {
            String msg = "Port must be a number between 1 and 65535";
            Log.d("[NET]", msg);
            DefaultActivity.toast(msg);
            return;
        }

        THREAD = new WirelessController(ipAddr, port);
        THREAD.start();
    }

    @Override
    public void run() {
        try {
            SOCKET = new Socket(this.address, this.port);
            WirelessSender.init(SOCKET.getOutputStream());
            new WirelessReceiver(SOCKET.getInputStream()).start();

        } catch (IOException e) {
            //TODO Handle THIS
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/me/knighthat/interactivedeck/connection/wireless/WirelessController.kt
package me.knighthat.interactivedeck.connection.wireless

import android.os.Build
import com.google.gson.JsonObject
import me.knighthat.interactivedeck.connection.RequestHandler
import me.knighthat.interactivedeck.event.EventHandler
import me.knighthat.interactivedeck.persistent.Persistent
import me.knighthat.interactivedeck.vars.Settings
import me.knighthat.lib.connection.Connection
import me.knighthat.lib.connection.request.PairRequest
import me.knighthat.lib.connection.wireless.WirelessReceiver
import me.knighthat.lib.connection.wireless.WirelessSender
import me.knighthat.lib.logging.Log
import me.knighthat.lib.observable.Observer
import java.io.InputStream
import java.net.Socket
import java.net.SocketException

class WirelessController(
    private val ip: String,
    private val port: Int
) : Thread() {

    private fun sendPairRequest() {
        val json = JsonObject()
        json.addProperty("brand", Build.BRAND)
        json.addProperty("device", Build.DEVICE)
        json.addProperty("manufacturer", Build.MANUFACTURER)
        json.addProperty("model", Build.MODEL)
        json.addProperty("androidVersion", Build.VERSION.RELEASE)
        PairRequest(json).send()
    }

    private fun setupReceiver(inStream: InputStream) {
        name = "NET/I"

        runCatching {
            WirelessReceiver(
                inStream,
                Settings.BUFFER,
                RequestHandler()
            ).run()
        }.onFailure {
            if (it is SocketException && Connection.isDisconnected())
                return

            Log.exc("Parsing request failed!", it, true)
            Log.reportBug()
        }
    }

    private fun addConnectionWatcher(client: Socket) {
        Connection.whenConnectionStatusChanged(object : Observer<Connection.Status> {
            override fun update(oldValue: Connection.Status?, newValue: Connection.Status?) {
                // Skip this if connected is established
                if (Connection.isConnected())
                    return
                
                Persistent.free()

                val currentActivity = EventHandler.getCurrentActivity() ?: return
                currentActivity.finish()
                EventHandler.setCurrentActivity(null)

                if (Connection.isDisconnected()) {
                    Settings.saveLastHost(ip, port)
                    client.close()
                }

                this@WirelessController.interrupt()
            }
        })
    }

    override fun run() {
        Socket(ip, port)
            .runCatching {

                Connection.status = Connection.Status.CONNECTED

                val sender = WirelessSender(getOutputStream())
                sender.start()
                addConnectionWatcher(this)

                sendPairRequest()
                setupReceiver(getInputStream())

                name = "NET"
                Connection.status = Connection.Status.DISCONNECTED
                sender.interrupt()

                close()

            }.onFailure {

                Connection.status = Connection.Status.ERROR
                Log.exc("Connection error!", it, false)

            }
    }
}